### PR TITLE
Refactor the Export path to read from the engine

### PR DESCRIPTION
**Pull request by VinaiRachakonda**
This pr removes the need of TableDataLoc from export and moves it towards an engine based table reader. It also modifies the writer implementations to work with sql.Row

TODO: I don't think export should support either --continue or --mapping. They were never tested nor are they relevant. If there is an error reading a row from a table, you should fail hard.

### CODE REVIEW

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 1]
```diff
@@ -193,6 +191,7 @@ func (cmd ExportCmd) ArgParser() *argparser.ArgParser {
 	ap.ArgListHelp = append(ap.ArgListHelp, [2]string{"table", "The table being exported."})
 	ap.ArgListHelp = append(ap.ArgListHelp, [2]string{"file", "The file being output to."})
 	ap.SupportsFlag(forceParam, "f", "If data already exists in the destination, the force flag will allow the target to be overwritten.")
+	// TODO: We shoould not support this for export
 	ap.SupportsFlag(contOnErrParam, "", "Continue exporting when row export errors are encountered.")
```
Comment (VinaiRachakonda) [inline go/cmd/dolt/commands/tblcmds/export.go:195]: @andrew-wm-arthur I don't think we should support this. It doesn't make much sense for export and was never originally tested.

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 2]
```diff
@@ -193,6 +191,7 @@ func (cmd ExportCmd) ArgParser() *argparser.ArgParser {
...
 	ap.SupportsString(schemaParam, "s", "schema_file", "The schema for the output data.")
 	ap.SupportsString(mappingFileParam, "m", "mapping_file", "A file that lays out how fields should be mapped from input data to output data.")
```
Comment (VinaiRachakonda) [inline go/cmd/dolt/commands/tblcmds/export.go:197]: Same with the mapping file command. Not sure it has anything to do with the export path.

File: go/cmd/dolt/commands/tblcmds/export.go (hunk 1 above)
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/tblcmds/export.go:195]: we can drop it

File: go/cmd/dolt/commands/tblcmds/export.go (hunk 2 above)
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/tblcmds/export.go:197]: we can drop this too

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 3]
```diff
@@ -221,31 +220,32 @@ func (cmd ExportCmd) Exec(ctx context.Context, commandStr string, args []string,
...
+		return commands.HandleVErrAndExitCode(errhand.BuildDError("Error creating reader for %s.", exOpts.SrcName()).AddCause(err).Build(), usage)
+	}
 
+	wr, verr := getTableWriter(ctx, root, dEnv, rd.GetSchema(), exOpts, importStatsCB)
 	if verr != nil {
 		return commands.HandleVErrAndExitCode(verr, usage)
 	}
 
-	skipped, verr := mvdata.MoveData(ctx, dEnv, mover, exOpts)
+	skipped, err := export(ctx, rd, wr, root.VRW(), exOpts)
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:233]: seems like `skipped` and `exOpts` are unused?

File: go/libraries/doltcore/mvdata/engine_mover.go [hunk 4]
```diff
@@ -298,7 +298,7 @@ func (s *sqlEngineMover) getInsertNode(inputChannel chan sql.Row) (sql.Node, err
 	}
 }
 
-// createInsertImportNode creates the relevant/analyzed insert node given the import option. This insert node is wrapped
+// createInsertImportNode creates the relevant/analyzed insert iter given the import option. This insert iter is wrapped
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/mvdata/engine_mover.go:301]: intended?

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 5]
```diff
@@ -285,17 +269,83 @@ func NewExportDataMover(ctx context.Context, root *doltdb.RootValue, dEnv *env.D
...
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case parsedRowChan <- row:
+			}
+		}
+	})
+
+	g.Go(func() error {
+		defer wr.Close(ctx)
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:333]: unhandled error

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 6]
```diff
@@ -285,17 +269,83 @@ func NewExportDataMover(ctx context.Context, root *doltdb.RootValue, dEnv *env.D
...
+			if err != nil {
+				return err
+			}
+
+			return wr.WriteRow(ctx, dRow)
+		}
+	}
+	g.Go(func() error {
+		defer close(parsedRowChan)
+		defer rd.Close(ctx)
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:312]: unhandled error

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 7]
```diff
@@ -285,17 +269,83 @@ func NewExportDataMover(ctx context.Context, root *doltdb.RootValue, dEnv *env.D
...
-	emptyTransColl := pipeline.NewTransformCollection()
+	return wr, nil
+}
+
+func export(ctx context.Context, rd table.TableReadCloser, wr table.TableWriteCloser, vrw types.ValueReadWriter, exOpts *exportOptions) (int64, error) {
+	g, ctx := errgroup.WithContext(ctx)
+
+	parsedRowChan := make(chan sql.Row)
+
+	getNextRow := func(rd table.TableReadCloser) (sql.Row, error) {
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:285]: `getNextRow` and `writeRow` make more sense as helper functions than as closures

**Review by andrew-wm-arthur — APPROVED** [review]
generally looks good, there's some room for cleanup

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 8]
```diff
@@ -277,42 +249,51 @@ func getTableWriter(ctx context.Context, root *doltdb.RootValue, dEnv *env.DoltE
...
+
+	parsedRowChan := make(chan sql.Row)
+
+	getNextRow := getNextRowFunc(rd)
+
 	g.Go(func() error {
 		defer close(parsedRowChan)
-		defer rd.Close(ctx)
+		defer func() {
+			err = rd.Close(ctx)
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:292]:
```suggestion
			if cerr = rd.Close(ctx); cerr != nil {
				err = cerr
			}
```

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 9]
```diff
@@ -330,14 +311,16 @@ func export(ctx context.Context, rd table.TableReadCloser, wr table.TableWriteCl
 	})
 
 	g.Go(func() error {
-		defer wr.Close(ctx)
+		defer func() {
+			err = wr.Close(ctx)
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:315]:
```suggestion
			if cerr = wr.Close(ctx); cerr != nil {
				err = cerr
			}
```

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 10]
```diff
@@ -277,42 +249,51 @@ func getTableWriter(ctx context.Context, root *doltdb.RootValue, dEnv *env.DoltE
...
+}
+
+func export(ctx context.Context, rd table.TableReadCloser, wr table.TableWriteCloser, vrw types.ValueReadWriter) (err error) {
+	g, ctx := errgroup.WithContext(ctx)
+
+	parsedRowChan := make(chan sql.Row)
+
+	getNextRow := getNextRowFunc(rd)
+
 	g.Go(func() error {
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:289]:
```suggestion
	g.Go(func() (err error) {
```

File: go/cmd/dolt/commands/tblcmds/export.go [hunk 11]
```diff
@@ -330,14 +311,16 @@ func export(ctx context.Context, rd table.TableReadCloser, wr table.TableWriteCl
 	})
 
 	g.Go(func() error {
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/tblcmds/export.go:313]:
```suggestion
	g.Go(func() (err error) {
```

**Review by andrew-wm-arthur — APPROVED** [review]
Couple of error handling improvements

**Pull request merged**